### PR TITLE
Update Numba to 0.57

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Array tests are currently broken

### CODE REVIEW

**Comment by guilhermeleobas** [comment]
Pending https://github.com/conda-forge/numba-feedstock/pull/115

**Comment by guilhermeleobas** [comment]
Blocked on HeavyDB 7.0 release on conda-forge

File: rbc/irtools.py [hunk 1]
```diff
@@ -438,6 +439,27 @@ def compile_to_LLVM(functions_and_signatures,
...
+               'Please, downgrade to a previous release of Numba that uses the '
+               'same LLVM version as the HeavyDB server. For more information, '
+               'see the table below:\n\n'
+               'https://github.com/numba/llvmlite#compatibility\n'
+               'https://github.com/heavyai/heavydb#dependencies\n\n'
+               f'To disable this error, run RBC with {flag}=1 flag enabled.')
+
+        DISABLE_LLVM_MISMATCH_ERROR = int(os.environ.get(flag, False))
+        if not DISABLE_LLVM_MISMATCH_ERROR:
+            raise LLVMVersionMismatchError(msg)
```
Comment (guilhermeleobas) [inline rbc/irtools.py:461]: Would this be the most appropriated place to put this check?

**Review by pearu — CHANGES_REQUESTED** [review]

File: .github/workflows/rbc_test.yml [hunk 2]
```diff
@@ -108,11 +108,25 @@ jobs:
       fail-fast: false
       matrix:
         os: [ubuntu-latest]
-        python-version: ['3.10', '3.9', '3.8']
-        numba-version: ['0.56', '0.55']
+        python-version: ['3.10', '3.9']
+        numba-version: ['0.56']
         heavydb-version: ['6.2', '6.1', '6.0']
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply .github/workflows/rbc_test.yml:113]: HeavyDB 6.4 has a bug where it crashes when executing UDFs/UDTFs.

File: environment.yml [hunk 3]
```diff
@@ -14,8 +14,8 @@ dependencies:
   - setuptools
   - numba>=0.55
   - llvmlite
-  - clang==11.1
-  - clangxx==11.1
+  - clang
+  - clangxx
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply environment.yml:18]: Depends on the environment. On CI? Yes! In a user environment, it depends on how mamba will solve it. But `clang/clangxx` are optional dependencies.

File: rbc/irtools.py (hunk 1 above)
Comment (guilhermeleobas) [reply rbc/irtools.py:461]: Okay, I can change that. We only check for the LLVM 11/14 combination, which is known to be problematic.

File: rbc/heavydb/array.py [hunk 4]
```diff
@@ -45,12 +45,11 @@ def deepcopy(self, context, builder, val, retptr):
...
-        struct_load = builder.load(val, name='struct_load')
-        src = builder.extract_value(struct_load, 0, name='array_buff_ptr')
-        element_count = builder.extract_value(struct_load, 1, name='array_size')
-        is_null = builder.extract_value(struct_load, 2, name='array_is_null')
-
         zero, one, two = int32_t(0), int32_t(1), int32_t(2)
+        src = builder.load(builder.gep(val, [zero, zero]), name='array_buff_ptr')
+        element_count = builder.load(builder.gep(val, [zero, one]), name='array_size')
+        is_null = builder.load(builder.gep(val, [zero, two]), name='array_is_null')
+
```
Comment (pearu) [inline rbc/heavydb/array.py:52]: Could you create a separate PR for this fix because landing the PR is blocked by heavydb 7 release.

File: rbc/heavydb/buffer.py [hunk 5]
```diff
@@ -207,7 +208,10 @@ def heavydb_buffer_constructor(context, builder, sig, args):
     ptr = memalloc(context, builder, ptr_type, element_count, element_size)
 
     llty = context.get_value_type(sig.return_type.dtype)
-    st_ptr = builder.alloca(llty)
+    st_size = context.get_abi_sizeof(llty)
+    st_ptr = builder.bitcast(allocate_varlen_buffer(builder, int64_t(st_size),
+                                                    int64_t(1)),
+                             llty.as_pointer())
```
Comment (pearu) [inline rbc/heavydb/buffer.py:214]: Add this to the separate PR.

File: rbc/heavydb/text_encoding_none.py [hunk 6]
```diff
@@ -176,7 +176,10 @@ def text_encoding_none_unicode_ctor(context, builder, sig, args):
     uni_str = context.make_helper(builder, sig.args[0], value=args[0])
 
     llty = context.get_value_type(sig.return_type.dtype)
-    st_ptr = builder.alloca(llty)
+    st_size = context.get_abi_sizeof(llty)
+    st_ptr = builder.bitcast(allocate_varlen_buffer(builder, int64_t(st_size),
+                                                    int64_t(1)),
+                             llty.as_pointer())
     zero, one, two = i32(0), i32(1), i32(2)
```
Comment (pearu) [inline rbc/heavydb/text_encoding_none.py:183]: Add this to the separate PR.

**Review by pearu — CHANGES_REQUESTED** [review]
@guilhermeleobas since we don't know when heavydb 7 will be release (so merging this PR is blocked), could you move the `alloc` releated fixes to a separate PR that can be landed earlier?

File: doc/envvars.rst [hunk 7]
```diff
@@ -14,3 +14,9 @@ Debugging
 
     If set to non-zero, insert debug statements to our implementation
     of Numba Runtime (NRT)
+
+.. envvar:: DISABLE_LLVM_MISMATCH_WARN
```
Comment (pearu) [inline doc/envvars.rst:18]:
```suggestion
.. envvar:: RBC_DISABLE_LLVM_MISMATCH_WARN
```

File: rbc/heavydb/remoteheavydb.py [hunk 8]
```diff
@@ -1266,8 +1268,35 @@ def _make_udf(self, caller, orig_sig, sig):
...
+            target = self.targets['cpu']
+            server_llvm_version = target.llvm_version
+            client_llvm_version = llvm.llvm_version_info
+
+            if (server_llvm_version[0], client_llvm_version[0]) == (11, 14):
+                c_llvm = '.'.join(map(str, client_llvm_version))
+                s_llvm = '.'.join(map(str, server_llvm_version))
+                msg = (
+                    f'The client LLVM version ({c_llvm}) is greater than the server '
+                    f'LLVM version ({s_llvm}). This is known to be unsupported. '
```
Comment (pearu) [inline rbc/heavydb/remoteheavydb.py:1286]: With #551, I believe this statement has become out-of-date. Could you rephrase the message along the lines of `"If you experience any issues that could be due to the discrepancy of LLVM versions, please try to downgrade/update Numba that uses LLVM version ({s_llvm})."`. Also, this statement can be generalized to all LLVM version combinations, not just `(11, 14)`.

File: rbc/heavydb/remoteheavydb.py [hunk 9]
```diff
@@ -1266,8 +1268,35 @@ def _make_udf(self, caller, orig_sig, sig):
             name + sig.mangling(),
             atypes, rtype, annotations)
 
+    def _check_llvm_version(self):
+        # check LLVM version before compiling to LLVM
+        flag_name = 'RBC_DISABLE_LLVM_MISMATCH_WARN'
+        flag = int(os.environ.get(flag_name, False))
+
```
Comment (pearu) [inline rbc/heavydb/remoteheavydb.py:1275]: We should check the LLVM version and possibly warn only once. One possible approach is as follows:
```suggestion
        os.environ[flag_name] = 1
```

File: rbc/heavydb/remoteheavydb.py (hunk 8 above)
Comment (pearu) [reply rbc/heavydb/remoteheavydb.py:1286]: I now see that the purpose is to warn only about the problematic LLVM version combinations. Since 11 and 14 are not problematic anymore, this function is essentially a no-op? What do you think?

In fact, the combination of 14 and 11 might be more problematic as a newer client-side LLVM version may use features (such as LLVM attributes or similar) that the server LLVM IR parser cannot handle.
